File: backend/controllers/checkoutController.js
const stripe = require('stripe')(process.env.STRIPE_SECRET_KEY);
const currencyService = require('../services/currencyService');
const { toStripeAmount, calculatePlatformFee } = require('../utils/currencyUtils');
const admin = require('firebase-admin');
const db = admin.firestore();

exports.createCheckoutSession = async (req, res) => {
  try {
    const { 
      pageId, 
      priceId, 
      currency, 
      successUrl, 
      cancelUrl,
      customerEmail,
      metadata 
    } = req.body;
    
    // Récupérer les informations de la page
    const pageDoc = await db.collection('pages').doc(pageId).get();
    if (!pageDoc.exists) {
      return res.status(404).json({ error: 'Page not found' });
    }
    
    const pageData = pageDoc.data();
    
    // Configuration de la session Checkout
    const sessionConfig = {
      payment_method_types: ['card'],
      line_items: [{
        price: priceId,
        quantity: 1
      }],
      mode: 'subscription',
      success_url: successUrl || `${process.env.DOMAIN}/public/success.html?session_id={CHECKOUT_SESSION_ID}`,
      cancel_url: cancelUrl || `${process.env.DOMAIN}/public/cancel.html`,
      customer_email: customerEmail,
      metadata: {
        pageId,
        currency,
        ...metadata
      },
      subscription_data: {
        metadata: {
          pageId,
          currency
        }
      }
    };
    
    // Si Stripe Connect est configuré pour cette page
    if (pageData.stripeAccountId) {
      const amount = await this.getPriceAmount(priceId);
      const platformFee = calculatePlatformFee(amount);
      
      sessionConfig.payment_intent_data = {
        application_fee_amount: toStripeAmount(platformFee, currency),
        transfer_data: {
          destination: pageData.stripeAccountId
        }
      };
    }
    
    // Créer la session
    const session = await stripe.checkout.sessions.create(sessionConfig);
    
    // Enregistrer la tentative de checkout
    await db.collection('checkout_attempts').add({
      sessionId: session.id,
      pageId,
      currency,
      status: 'pending',
      createdAt: new Date(),
      metadata
    });
    
    res.json({
      id: session.id,
      url: session.url
    });
    
  } catch (error) {
    console.error('Checkout session error:', error);
    res.status(500).json({ error: error.message });
  }
};

exports.createPricesForAllCurrencies = async (req, res) => {
  try {
    const { productName, basePriceUSD, interval = 'month' } = req.body;
    
    // Créer le produit Stripe
    const product = await stripe.products.create({
      name: productName,
      metadata: {
        createdAt: new Date().toISOString()
      }
    });
    
    // Créer les prix pour toutes les devises
    const prices = await currencyService.createMultiCurrencyPrices(
      product.id,
      basePriceUSD,
      interval
    );
    
    // Sauvegarder dans Firebase
    await db.collection('products').doc(product.id).set({
      productId: product.id,
      name: productName,
      basePriceUSD,
      interval,
      prices,
      createdAt: new Date()
    });
    
    res.json({
      productId: product.id,
      prices
    });
    
  } catch (error) {
    console.error('Create prices error:', error);
    res.status(500).json({ error: error.message });
  }
};

exports.getPrices = async (req, res) => {
  try {
    const { productId } = req.params;
    const { currency } = req.query;
    
    // Récupérer depuis Firebase
    const productDoc = await db.collection('products').doc(productId).get();
    
    if (!productDoc.exists) {
      return res.status(404).json({ error: 'Product not found' });
    }
    
    const productData = productDoc.data();
    
    if (currency) {
      // Retourner le prix pour une devise spécifique
      const price = productData.prices[currency];
      if (!price) {
        return res.status(404).json({ error: 'Price not found for this currency' });
      }
      res.json(price);
    } else {
      // Retourner tous les prix
      res.json(productData.prices);
    }
    
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

exports.createPortalSession = async (req, res) => {
  try {
    const { customerId } = req.body;
    
    const session = await stripe.billingPortal.sessions.create({
      customer: customerId,
      return_url: `${process.env.DOMAIN}/public/dashboard.html`
    });
    
    res.json({ url: session.url });
    
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

exports.handleWebhook = async (req, res) => {
  const sig = req.headers['stripe-signature'];
  let event;
  
  try {
    event = stripe.webhooks.constructEvent(
      req.body,
      sig,
      process.env.STRIPE_WEBHOOK_SECRET
    );
  } catch (err) {
    console.error('Webhook signature verification failed:', err);
    return res.status(400).send(`Webhook Error: ${err.message}`);
  }
  
  // Gérer les différents types d'événements
  switch (event.type) {
    case 'checkout.session.completed':
      await this.handleCheckoutComplete(event.data.object);
      break;
      
    case 'customer.subscription.created':
      await this.handleSubscriptionCreated(event.data.object);
      break;
      
    case 'customer.subscription.updated':
      await this.handleSubscriptionUpdated(event.data.object);
      break;
      
    case 'customer.subscription.deleted':
      await this.handleSubscriptionDeleted(event.data.object);
      break;
      
    case 'invoice.payment_succeeded':
      await this.handlePaymentSucceeded(event.data.object);
      break;
      
    case 'invoice.payment_failed':
      await this.handlePaymentFailed(event.data.object);
      break;
      
    default:
      console.log(`Unhandled event type ${event.type}`);
  }
  
  res.json({ received: true });
};

// Méthodes privées pour gérer les événements
exports.handleCheckoutComplete = async (session) => {
  try {
    const { pageId, currency } = session.metadata;
    
    // Mettre à jour le statut du checkout
    const checkoutQuery = await db.collection('checkout_attempts')
      .where('sessionId', '==', session.id)
      .limit(1)
      .get();
    
    if (!checkoutQuery.empty) {
      const checkoutDoc = checkoutQuery.docs[0];
      await checkoutDoc.ref.update({
        status: 'completed',
        completedAt: new Date(),
        customerId: session.customer,
        subscriptionId: session.subscription
      });
    }
    
    // Créer l'enregistrement de l'abonnement
    await db.collection('subscriptions').add({
      userId: session.client_reference_id,
      pageId,
      stripeCustomerId: session.customer,
      stripeSubscriptionId: session.subscription,
      currency,
      status: 'active',
      createdAt: new Date()
    });
    
    console.log('✅ Checkout completed:', session.id);
    
  } catch (error) {
    console.error('Error handling checkout complete:', error);
  }
};

exports.handleSubscriptionCreated = async (subscription) => {
  try {
    const { pageId, currency } = subscription.metadata;
    
    // Mettre à jour ou créer l'enregistrement
    await db.collection('subscriptions').doc(subscription.id).set({
      stripeSubscriptionId: subscription.id,
      stripeCustomerId: subscription.customer,
      pageId,
      currency,
      status: subscription.status,
      currentPeriodStart: new Date(subscription.current_period_start * 1000),
      currentPeriodEnd: new Date(subscription.current_period_end * 1000),
      cancelAtPeriodEnd: subscription.cancel_at_period_end,
      createdAt: new Date()
    }, { merge: true });
    
    console.log('✅ Subscription created:', subscription.id);
    
  } catch (error) {
    console.error('Error handling subscription created:', error);
  }
};

// Helper pour obtenir le montant d'un prix
exports.getPriceAmount = async (priceId) => {
  try {
    const price = await stripe.prices.retrieve(priceId);
    return price.unit_amount / 100; // Convertir depuis les centimes
  } catch (error) {
    console.error('Error retrieving price:', error);
    return 0;
  }
};

File: telegram/controllers/CheckoutController.py
import stripe
import os
from datetime import datetime
from services.CheckoutService import CheckoutService
from services.CurrencyService import CurrencyService
from config.database import db
import json

class CheckoutController:
    def __init__(self):
        self.checkout_service = CheckoutService()
        self.currency_service = CurrencyService()
        stripe.api_key = os.getenv('STRIPE_SECRET_KEY')
    
    def create_checkout_session(self, request):
        """Créer une session de paiement Stripe"""
        try:
            data = request.get_json()
            
            # Données requises
            page_id = data.get('pageId')
            price_id = data.get('priceId')
            currency = data.get('currency', 'USD')
            success_url = data.get('successUrl', os.getenv('SUCCESS_URL'))
            cancel_url = data.get('cancelUrl', os.getenv('CANCEL_URL'))
            customer_email = data.get('customerEmail')
            metadata = data.get('metadata', {})
            
            # Récupérer les infos de la page
            page_ref = db.collection('pages').document(page_id)
            page = page_ref.get()
            
            if not page.exists:
                return {'error': 'Page not found'}, 404
            
            page_data = page.to_dict()
            
            # Configuration session
            session_config = {
                'payment_method_types': ['card'],
                'line_items': [{
                    'price': price_id,
                    'quantity': 1
                }],
                'mode': 'subscription',
                'success_url': f"{success_url}?session_id={{CHECKOUT_SESSION_ID}}",
                'cancel_url': cancel_url,
                'customer_email': customer_email,
                'metadata': {
                    'pageId': page_id,
                    'currency': currency,
                    **metadata
                },
                'subscription_data': {
                    'metadata': {
                        'pageId': page_id,
                        'currency': currency
                    }
                }
            }
            
            # Si Stripe Connect configuré
            if page_data.get('stripeAccountId'):
                amount = self.checkout_service.get_price_amount(price_id)
                platform_fee = self.currency_service.calculate_platform_fee(amount)
                
                session_config['payment_intent_data'] = {
                    'application_fee_amount': int(platform_fee * 100),
                    'transfer_data': {
                        'destination': page_data['stripeAccountId']
                    }
                }
            
            # Créer la session
            session = stripe.checkout.Session.create(**session_config)
            
            # Enregistrer la tentative
            db.collection('checkout_attempts').add({
                'sessionId': session.id,
                'pageId': page_id,
                'currency': currency,
                'status': 'pending',
                'createdAt': datetime.now(),
                'metadata': metadata
            })
            
            return {
                'id': session.id,
                'url': session.url
            }, 200
            
        except Exception as e:
            print(f"Checkout error: {e}")
            return {'error': str(e)}, 500
    
    def get_prices(self, product_id, request):
        """Obtenir les prix dans toutes les devises"""
        try:
            currency = request.args.get('currency')
            
            # Récupérer depuis Firebase
            product_ref = db.collection('products').document(product_id)
            product = product_ref.get()
            
            if not product.exists:
                return {'error': 'Product not found'}, 404
            
            product_data = product.to_dict()
            
            if currency:
                # Prix pour une devise spécifique
                price = product_data.get('prices', {}).get(currency)
                if not price:
                    return {'error': f'Price not found for {currency}'}, 404
                return price, 200
            else:
                # Tous les prix
                return product_data.get('prices', {}), 200
                
        except Exception as e:
            return {'error': str(e)}, 500
    
    def create_portal_session(self, request):
        """Créer session portail client"""
        try:
            data = request.get_json()
            customer_id = data.get('customerId')
            
            session = stripe.billing_portal.Session.create(
                customer=customer_id,
                return_url=f"{os.getenv('DOMAIN')}/public/dashboard.html"
            )
            
            return {'url': session.url}, 200
            
        except Exception as e:
            return {'error': str(e)}, 500
    
    def handle_webhook(self, event):
        """Gérer les webhooks Stripe"""
        try:
            # Router selon le type d'événement
            handlers = {
                'checkout.session.completed': self._handle_checkout_complete,
                'customer.subscription.created': self._handle_subscription_created,
                'customer.subscription.updated': self._handle_subscription_updated,
                'customer.subscription.deleted': self._handle_subscription_deleted,
                'invoice.payment_succeeded': self._handle_payment_succeeded,
                'invoice.payment_failed': self._handle_payment_failed
            }
            
            handler = handlers.get(event['type'])
            if handler:
                handler(event['data']['object'])
            else:
                print(f"Unhandled event type: {event['type']}")
            
            return {'received': True}, 200
            
        except Exception as e:
            print(f"Webhook error: {e}")
            return {'error': str(e)}, 500
    
    def _handle_checkout_complete(self, session):
        """Gérer checkout complété"""
        page_id = session['metadata'].get('pageId')
        currency = session['metadata'].get('currency')
        
        # Mettre à jour le statut
        attempts = db.collection('checkout_attempts')\
            .where('sessionId', '==', session['id'])\
            .limit(1)\
            .get()
        
        for attempt in attempts:
            attempt.reference.update({
                'status': 'completed',
                'completedAt': datetime.now(),
                'customerId': session['customer'],
                'subscriptionId': session['subscription']
            })
        
        # Créer l'abonnement
        db.collection('subscriptions').add({
            'userId': session.get('client_reference_id'),
            'pageId': page_id,
            'stripeCustomerId': session['customer'],
            'stripeSubscriptionId': session['subscription'],
            'currency': currency,
            'status': 'active',
            'createdAt': datetime.now()
        })
        
        print(f"✅ Checkout completed: {session['id']}")
    
    def _handle_subscription_created(self, subscription):
        """Gérer création abonnement"""
        page_id = subscription['metadata'].get('pageId')
        currency = subscription['metadata'].get('currency')
        
        db.collection('subscriptions').document(subscription['id']).set({
            'stripeSubscriptionId': subscription['id'],
            'stripeCustomerId': subscription['customer'],
            'pageId': page_id,
            'currency': currency,
            'status': subscription['status'],
            'currentPeriodStart': datetime.fromtimestamp(subscription['current_period_start']),
            'currentPeriodEnd': datetime.fromtimestamp(subscription['current_period_end']),
            'cancelAtPeriodEnd': subscription['cancel_at_period_end'],
            'createdAt': datetime.now()
        }, merge=True)
        
        print(f"✅ Subscription created: {subscription['id']}")